src/NLog/Layouts/Typed/Layout.cs

```diff
@@ -0,0 +1,315 @@
+        {
+            if (_type == null || raw == null)
+            {
+                value = default(T);
+                return false;
+            }
+
+            // We don't use DI here because of this will be called before DI setup and performance reasons
+            var cultureInfo = CultureInfo.CurrentCulture;
+            try
+            {
+                var convertedValue = ValueConverter.Instance.Convert(raw, _type, null, cultureInfo);
+                if (convertedValue is T goodValue)
+                {
+                    value = goodValue;
+                    return true;
+                }
+            }
+            catch (Exception e)
+            {
+                InternalLogger.Debug(e, "Conversion to type {0} failed", _type);
+                if (e.MustBeRethrown())
+                {
+                    throw;
+                }
+            }
+
+            value = default(T);
+            return false;
+        }
+
+        #endregion
+
+        #region Conversion
+
+        /// <summary>
+        /// Converts a given text to a <see cref="Layout" />.
+        /// </summary>
+        /// <param name="value">Text to be converted.</param>
+        /// <returns><see cref="SimpleLayout" /> object represented by the text.</returns>
+        public static implicit operator Layout<T>(T value)
+        {
+            return new Layout<T>(value);
+        }
+
+        /// <summary>
+        /// Converts a given text to a <see cref="Layout" />.
+        /// </summary>
+        /// <param name="layout">Text to be converted.</param>
+        /// <returns><see cref="SimpleLayout" /> object represented by the text.</returns>
+        public static implicit operator Layout<T>([Localizable(false)] string layout)
+        {
+            return new Layout<T>(layout);
+        }
+
+        #endregion
+
+        /// <inheritdoc />
+        protected override string GetFormattedMessage(LogEventInfo logEvent)
+        {
+            if (_fixedValue)
+            {
+                if (_value == null)
+                {
+                    return null;
+                }
+
+                var text = ValueToString(_value, LoggingConfiguration.DefaultCultureInfo);
+                if (text != null)
+                {
+                    return text;
+                }
+            }
+
+            return _layout.Render(logEvent);
+        }
+
+        /// <inheritdoc cref="IRawValue" />
+        internal override bool TryGetRawValue(LogEventInfo logEvent, out object rawValue)
+        {
+            if (_fixedValue)
+            {
+                rawValue = _value;
+                return true;
+            }
+
+            if (_layout == null)
+            {
+                rawValue = null;
+                return true;
+            }
+
+            if (_layout.TryGetRawValue(logEvent, out var raw))
+            {
+                var success = TryConvertRawToValue(raw, out var i);
+                rawValue = i;
```

Comment:
Funny that value is also assigned when success = false.

Reply:
Well we have to assign the out,. I assume here TryConvertRawToValue assigns the out correctly. Do you think this is wrong?

Reply:
Thinking something like this:

```
if (_layout.TryGetRawValue(logEvent, out var raw) && TryConvertRawToValue(raw, out var i))
{
    rawValue = i;
    return true;
}
```

Btw. might want to fix the variable name `i`.